File: klinara-ios/klinara-ios/Services/Auth/LiveAuthService.swift
import Foundation

/// ``AuthService``'in gerçek uygulaması — Faz 1 uçlarına konuşur.
///
/// Protokol değişmediği için ekranlarda tek satır değişmez; mock ile canlı
/// arasındaki fark yalnız ``ServiceContainer``'da hangi örneğin kurulduğudur.
struct LiveAuthService: AuthService {

    /// ``ServiceContainer`` oturum düşme geri çağrısını bağlayabilsin diye görünür.
    let client: APIClient

    init(client: APIClient) {
        self.client = client
    }

    // MARK: - Giriş

    nonisolated private struct LoginBody: Encodable, Sendable {
        let email: String?
        let phone: String?
        let password: String
        let deviceLabel: String
    }

    func login(
        email: String?,
        phone: String?,
        password: String,
        deviceLabel: String
    ) async throws -> LoginResponse {
        var request = APIRequest.post(
            "auth/login",
            body: LoginBody(email: email, phone: phone, password: password, deviceLabel: deviceLabel)
        )
        request.requiresAuth = false
        return try await client.send(request)
    }

    nonisolated private struct SelectTenantBody: Encodable, Sendable {
        let challengeToken: String
        let tenantId: String
    }

    func selectTenant(challengeToken: String, tenantId: String) async throws -> LoginResponse {
        var request = APIRequest.post(
            "auth/tenant",
            body: SelectTenantBody(challengeToken: challengeToken, tenantId: tenantId)
        )
        request.requiresAuth = false
        return try await client.send(request)
    }

    nonisolated private struct VerifyMfaBody: Encodable, Sendable {
        let challengeToken: String
        let code: String
        let deviceLabel: String
    }

    func verifyMfa(
        challengeToken: String,
        code: String,
        deviceLabel: String
    ) async throws -> LoginResponse {
        var request = APIRequest.post(
            "auth/2fa/verify",
            body: VerifyMfaBody(challengeToken: challengeToken, code: code, deviceLabel: deviceLabel)
        )
        request.requiresAuth = false
        return try await client.send(request)
    }

    // MARK: - TOTP kurulumu
    //
    // `/setup` ve `/enable` challenge token'ını GÖVDEDE değil `Authorization`
    // başlığında bekler (`totp.controller.ts` → `actorOf(request)`).

    func totpSetup(challengeToken: String) async throws -> TotpSetup {
        var request = APIRequest.post("auth/2fa/setup")
        request.bearerOverride = challengeToken
        return try await client.send(request)
    }

    nonisolated private struct TotpCodeBody: Encodable, Sendable {
        let code: String
    }

    nonisolated private struct BackupCodesResponse: Decodable, Sendable {
        let backupCodes: [String]
    }

    func totpEnable(challengeToken: String, code: String) async throws -> [String] {
        var request = APIRequest.post("auth/2fa/enable", body: TotpCodeBody(code: code))
        request.bearerOverride = challengeToken
        let response: BackupCodesResponse = try await client.send(request)
        return response.backupCodes
    }

    // MARK: - Passkey

    func passkeyAssertionOptions() async throws -> PasskeyOptions {
        var request = APIRequest.post("auth/passkey/options", body: EmptyBody())
        request.requiresAuth = false
        let raw: WebAuthnOptionsPayload = try await client.send(request)
        return try raw.assertionOptions()
    }

    func passkeyVerify(
        _ assertion: PasskeyAssertion,
        deviceLabel: String
    ) async throws -> LoginResponse {
        var request = APIRequest.post(
            "auth/passkey/verify",
            body: WebAuthnEnvelope(response: assertion.credentialJSON, deviceLabel: deviceLabel)
        )
        request.requiresAuth = false
        return try await client.send(request)
    }

    func passkeyRegistrationOptions() async throws -> PasskeyOptions {
        let raw: WebAuthnOptionsPayload = try await client.send(
            APIRequest.post("auth/passkeys/register/options")
        )
        return try raw.registrationOptions()
    }

    func registerPasskey(_ registration: PasskeyRegistration, deviceLabel: String) async throws {
        try await client.send(APIRequest.post(
            "auth/passkeys/register",
            body: WebAuthnEnvelope(response: registration.credentialJSON, deviceLabel: deviceLabel)
        ))
    }

    // MARK: - Oturum sonrası

    func me() async throws -> MeResponse {
        try await client.send(APIRequest.get("me"))
    }

    func branches() async throws -> [BranchSummary] {
        let response: ListEnvelope<BranchSummary> = try await client.send(APIRequest.get("branches"))
        return response.data
    }

    // MARK: - Telefon doğrulama

    nonisolated private struct StartPhoneBody: Encodable, Sendable {
        let phone: String
    }

    func startPhoneVerification(phone: String) async throws -> PhoneVerificationStarted {
        try await client.send(APIRequest.post("auth/phone/start", body: StartPhoneBody(phone: phone)))
    }

    nonisolated private struct VerifyPhoneBody: Encodable, Sendable {
        let code: String
    }

    func verifyPhone(code: String) async throws -> PhoneVerified {
        try await client.send(APIRequest.post("auth/phone/verify", body: VerifyPhoneBody(code: code)))
    }

    // MARK: - Parola kurtarma

    nonisolated private struct ForgotPasswordBody: Encodable, Sendable {
        let email: String
    }

    func forgotPassword(email: String) async throws {
        var request = APIRequest.post("auth/password/forgot", body: ForgotPasswordBody(email: email))
        request.requiresAuth = false
        try await client.send(request)
    }

    // MARK: - Oturum

    nonisolated private struct RefreshBody: Encodable, Sendable {
        let refreshToken: String
    }

    func refresh(refreshToken: String) async throws -> AuthTokens {
        var request = APIRequest.post("auth/refresh", body: RefreshBody(refreshToken: refreshToken))
        request.requiresAuth = false
        return try await client.send(request)
    }

    func logout() async throws {
        try await client.send(APIRequest.post("auth/logout"))
    }
}

private nonisolated struct EmptyBody: Encodable, Sendable {}

File: klinara-android/app/src/main/java/com/klinara/android/services/auth/LiveAuthService.kt
package com.klinara.android.services.auth

import com.klinara.android.services.networking.ApiClient
import com.klinara.android.services.networking.ApiRequest
import com.klinara.android.services.networking.KlinaraJson
import com.klinara.android.services.networking.ListEnvelope
import com.klinara.android.services.networking.RequestBodyPayload
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put

@Serializable
private data class LoginBody(
    val email: String? = null,
    val phone: String? = null,
    val password: String,
    val deviceLabel: String,
)

@Serializable
private data class SelectTenantBody(
    val challengeToken: String,
    val tenantId: String,
)

@Serializable
private data class VerifyMfaBody(
    val challengeToken: String,
    val code: String,
    val deviceLabel: String,
)

@Serializable
private data class TotpCodeBody(val code: String)

@Serializable
private data class BackupCodesResponse(val backupCodes: List<String> = emptyList())

@Serializable
private data class StartPhoneBody(val phone: String)

@Serializable
private data class VerifyPhoneBody(val code: String)

@Serializable
private data class ForgotPasswordBody(val email: String)

/** `{ response: <webauthn json>, deviceLabel }` — RegisterPasskeyDto / VerifyPasskeyDto. */
private fun webAuthnEnvelope(
    responseJson: String,
    deviceLabel: String,
): RequestBodyPayload {
    val response = KlinaraJson.parseToJsonElement(responseJson).jsonObject
    val envelope: JsonObject =
        buildJsonObject {
            put("response", response)
            put("deviceLabel", deviceLabel)
        }
    return RequestBodyPayload(KlinaraJson.encodeToString(JsonObject.serializer(), envelope))
}

private fun body(json: String) = RequestBodyPayload(json)

class LiveAuthService internal constructor(
    private val client: ApiClient,
) : AuthService {
    override suspend fun login(
        email: String?,
        phone: String?,
        password: String,
        deviceLabel: String,
    ): LoginResponse =
        client.send(
            ApiRequest.post(
                "auth/login",
                body = body(KlinaraJson.encodeToString(LoginBody(email, phone, password, deviceLabel))),
                requiresAuth = false,
            ),
        )

    override suspend fun selectTenant(
        challengeToken: String,
        tenantId: String,
    ): LoginResponse =
        client.send(
            ApiRequest.post(
                "auth/tenant",
                body = body(KlinaraJson.encodeToString(SelectTenantBody(challengeToken, tenantId))),
                requiresAuth = false,
            ),
        )

    override suspend fun verifyMfa(
        challengeToken: String,
        code: String,
        deviceLabel: String,
    ): LoginResponse =
        client.send(
            ApiRequest.post(
                "auth/2fa/verify",
                body = body(KlinaraJson.encodeToString(VerifyMfaBody(challengeToken, code, deviceLabel))),
                requiresAuth = false,
            ),
        )

    override suspend fun totpSetup(challengeToken: String): TotpSetup =
        client.send(
            ApiRequest.post("auth/2fa/setup", requiresAuth = false, bearerOverride = challengeToken),
        )

    override suspend fun totpEnable(
        challengeToken: String,
        code: String,
    ): List<String> =
        client
            .send<BackupCodesResponse>(
                ApiRequest.post(
                    "auth/2fa/enable",
                    body = body(KlinaraJson.encodeToString(TotpCodeBody(code))),
                    requiresAuth = false,
                    bearerOverride = challengeToken,
                ),
            ).backupCodes

    override suspend fun me(): MeResponse = client.send(ApiRequest.get("me"))

    override suspend fun branches(): List<BranchSummary> =
        client.send<ListEnvelope<BranchSummary>>(ApiRequest.get("branches")).data

    override suspend fun startPhoneVerification(phone: String): PhoneVerificationStarted =
        client.send(
            ApiRequest.post("auth/phone/start", body = body(KlinaraJson.encodeToString(StartPhoneBody(phone)))),
        )

    override suspend fun verifyPhone(code: String): PhoneVerified =
        client.send(
            ApiRequest.post("auth/phone/verify", body = body(KlinaraJson.encodeToString(VerifyPhoneBody(code)))),
        )

    override suspend fun forgotPassword(email: String) {
        client.sendVoid(
            ApiRequest.post(
                "auth/password/forgot",
                body = body(KlinaraJson.encodeToString(ForgotPasswordBody(email))),
                requiresAuth = false,
            ),
        )
    }

    override suspend fun logout() {
        client.sendVoid(ApiRequest.post("auth/logout"))
    }

    override suspend fun totpStatus(): TotpStatus = client.send(ApiRequest.get("auth/2fa"))

    override suspend fun passkeys(): List<PasskeySummary> =
        client.send<ListEnvelope<PasskeySummary>>(ApiRequest.get("auth/passkeys")).data

    override suspend fun deletePasskey(id: String) {
        client.sendVoid(ApiRequest.delete("auth/passkeys/$id"))
    }

    override suspend fun passkeyAssertionOptions(): String =
        client
            .send<JsonObject>(ApiRequest.post("auth/passkey/options", requiresAuth = false))
            .let { KlinaraJson.encodeToString(JsonObject.serializer(), it) }

    override suspend fun passkeyVerify(
        responseJson: String,
        deviceLabel: String,
    ): LoginResponse =
        client.send(
            ApiRequest.post(
                "auth/passkey/verify",
                body = webAuthnEnvelope(responseJson, deviceLabel),
                requiresAuth = false,
            ),
        )

    override suspend fun passkeyRegistrationOptions(): String =
        client
            .send<JsonObject>(ApiRequest.post("auth/passkeys/register/options"))
            .let { KlinaraJson.encodeToString(JsonObject.serializer(), it) }

    override suspend fun registerPasskey(
        responseJson: String,
        deviceLabel: String,
    ) {
        client.sendVoid(
            ApiRequest.post("auth/passkeys/register", body = webAuthnEnvelope(responseJson, deviceLabel)),
        )
    }
}
